gh-140344: ast: Add deprecation warnings

jepler opened the pull request:
<!-- gh-issue-number: gh-140344 -->
* Issue: gh-140344
<!-- /gh-issue-number -->

StanFromIreland reviewed (dismissed):
Please see Hugo's comment on the issue, these need to raise deprecation warnings till 3.17 minimum (I don't think we need to do five years here, two should be fine).

jepler committed:
gh-140344: ast: Add deprecation warnings

These were all deprecated in 3.9 (bace59d8b8) but without
a runtime deprecation warning. Add it now, so that these
items can be removed in 3.20 per PEP 387.

JelleZijlstra commented on Misc/NEWS.d/next/Library/2025-10-20-09-12-18.gh-issue-140344.WjbYg-.rst:
```diff
@@ -0,0 +1,3 @@
+The classes ``ast.Suite``, ``ast.AugLoad``, ``ast.AugStore``, and
```
This only lists four of the seven things deprecated. It also doesn't mention the `.dims` attribute.

jepler replied:
added, thanks.

JelleZijlstra commented on Lib/ast.py:
```diff
@@ -689,5 +692,23 @@ def main(args=None):
                indent=args.indent, show_empty=args.show_empty))
 
+_deprecated = {
+        'slice': globals().pop("slice"),
```
Note this means `from ast import *` will no longer pick these up. That's probably OK for 3.15 but let's mention it in docs.

itamaro commented on Misc/NEWS.d/next/Library/2025-10-20-09-12-18.gh-issue-140344.WjbYg-.rst:
```diff
@@ -0,0 +1,3 @@
+The classes ``ast.Suite``, ``ast.AugLoad``, ``ast.AugStore``, and
```
Any migration guidance if anyone still uses these?

jepler replied:
You can find discussion around the original deprecations, e.g., https://github.com/python/cpython/issues/84169 but I didn't find any specific migration advice.

An ast produced by `ast.parse` will never include instances of these classes.

The implementation of all the removed classes are pretty trivial so if you need them for some reason you can probably just copy them. e.g., AugLoad has no body beyond a docstring.

```
  class AugLoad(expr_context):
      """Deprecated AST node class.  Unused in Python 3."""
```

JelleZijlstra replied:
Yeah there's no good use for them. Most likely if people are using them, it's just some code that does something for all AST nodes, and they should simply delete the code handling these nodes. But that's guesswork.

jepler committed:
NEWS: Finish detailing the change to the ast module.

jepler committed:
Doc: further document ast deprecations & removals

JelleZijlstra commented on Doc/deprecations/pending-removal-in-3.20.rst:
```diff
@@ -22,3 +22,14 @@ Pending removal in Python 3.20
 
   (Contributed by Hugo van Kemenade and Stan Ulbrych in :gh:`76007`.)
+
+* The ``__version__`` attribute has been deprecated in these standard library
```
This is duplicated.

StanFromIreland commented on Lib/ast.py:
```diff
@@ -22,7 +22,6 @@
 """
 from _ast import *
 
-
```
Please revert unrelated changes.

JelleZijlstra committed:
Merge remote-tracking branch 'upstream/main' into pr-140345-jepler-gh-140344

# Conflicts:
#	Doc/deprecations/pending-removal-in-3.20.rst
#	Doc/whatsnew/3.15.rst
#	Lib/test/test_ast/test_ast.py

JelleZijlstra committed:
tweak docs

JelleZijlstra committed:
tweak more

JelleZijlstra committed:
ignore

jepler committed:
Merge remote-tracking branch 'origin/main' into gh-140344

jepler committed:
gh-140344: Re-target deprecation to occur in 3.21

jepler committed:
remove unintended change to file

jepler committed:
fix version number in header

jepler committed:
ast: use lazy-import for warnings module

jepler committed:
add 3.21 pending deprecations to index
